scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,355 @@
+            .apply(GroupByKey.create())
+            .apply("Write to Elasticesarch",
+                   ParDo.of(new ElasticsearchWriter<>
+                                (clusterName, servers, maxBulkRequestSize, toDocWriteRequests,
+                                 error)));
+        return PDone.in(input.getPipeline());
+      }
+    }
+
+    private static class AssignToShard<T> extends DoFn<T, KV<Long, T>> {
+
+      private final long numOfShard;
+
+      public AssignToShard(long numOfShard) {
+        this.numOfShard = numOfShard;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        // assign this element to a random shard
+        final long shard = ThreadLocalRandom.current().nextLong(numOfShard);
+        c.output(KV.of(shard, c.element()));
+      }
+    }
+
+    private static class ElasticsearchWriter<T> extends DoFn<KV<Long, Iterable<T>>, Void> {
+
+      private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchWriter.class);
+      private final ClientSupplier clientSupplier;
+      private final SerializableFunction<T, Iterable<DocWriteRequest<?>>> toDocWriteRequests;
+      private final ThrowingConsumer<BulkExecutionException> error;
+      private final int maxBulkRequestSize;
+
+      public ElasticsearchWriter(String clusterName,
+                                 InetSocketAddress[] servers,
+                                 int maxBulkRequestSize,
+                                 SerializableFunction<T, Iterable<DocWriteRequest<?>>> toDocWriteRequests,
+                                 ThrowingConsumer<BulkExecutionException> error) {
+        this.maxBulkRequestSize = maxBulkRequestSize;
+        this.clientSupplier = new ClientSupplier(clusterName, servers);
+        this.toDocWriteRequests = toDocWriteRequests;
+        this.error = error;
+      }
+
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        final Iterable<T> values = c.element().getValue();
+
+        // Elasticsearch throws ActionRequestValidationException if bulk request is empty,
+        // so do nothing if number of actions is zero.
+        if (!values.iterator().hasNext()) {
+          LOG.info("ElasticsearchWriter: no requests to send");
+          return;
+        }
+
+        final Stream<DocWriteRequest> docWriteRequests =
+            StreamSupport.stream(values.spliterator(), false)
+                .map(toDocWriteRequests::apply)
+                .flatMap(ar -> StreamSupport.stream(ar.spliterator(), false));
+
+        final Iterable<List<DocWriteRequest>> chunks =
+            Iterables.partition(docWriteRequests::iterator, maxBulkRequestSize);
+        chunks.forEach(chunk -> {
+          try {
+            final BulkRequest bulkRequest = new BulkRequest().add(chunk.toArray(
+                new DocWriteRequest[0]));
+            final BulkResponse bulkItemResponse = clientSupplier.get().bulk(bulkRequest).get();
```

Comment:
depending on how this worker is threaded, would it make more sense to use java Future callbacks instead of blocking for every request?

Reply:
Threaded model here is managed by beam and in this case since it's a sink it's kinda "ok" to block, since each list of bulk requests happen in a different thread and we don't have a proper rate-limit/retry police in place.